Streamline tridiagonalisation of JacobianLinearOperator/FunctionLinearOperator using coloring method

jpbrodrick89 opened the pull request:
BREAKING CHANGE: Tridiagonals are no longer "extracted" from operators but rely on the promise of the `tridiagonal` tag being accurate and fulfilled. If this promise is broken solutions may differ from user expectations.

### Preface

Very similar in spirit to #164 and partially addresses the problem discussed in #149. This turned out to be a lot simpler than I anticipated and thought it would be helpful to compare and contrast the two implementations and performance impacts. In this case, the performance impact is significant and consistent for array sizes > 75 even when the JVP batching rule is efficient. However, this implementation does not support primitives with no batching rule is available at all. As I am just `vmap`ping over three vectors perhaps I could write an explicit `for` loop but probably best to prioritise efficiency here over flexibility here (or maybe we can do both if we can query whether the rule exists within a jit function?).

Some of the variable names and overall spirit of the method is inspired direct from [sparsejac](https://github.com/mfschubert/sparsejac) so we should probably add attribution (under MIT license) somewhere.

### Changes made

The single-dispatch `tridiagonal` function has been re-written for `JacobianLinearOperator`/`FunctionLinearOperator` based on the observation made by coloring methods that all elements of a tridiagonal matrix can be disentagled from just three carefully constructed vector pre/post-multiplications. I have gone with the simple case where each vector consists of ones in every third position and zeros otherwise. This means the full Jacobian need never be calculated.

### Performance Impact

Applying this to the problem discussed in #149 where we attempt to use `optimistix` to provide an implicit solve of a diffusion equation and compare this with straightforward `lineax.linear_solve`s with various operators (with `EQX_ON_ERROR=nan`):

<details>
<summary>
Code
</summary>

```python
import jax.numpy as jnp
import jax
import lineax
import optimistix
import sparsejac
jax.config.update("jax_enable_x64", True)
#jax.config.update("jax_disable_jit", True)

key = jax.random.PRNGKey(0)

def laplacian_numerator(y):
    return jnp.diff(jnp.diff(y), prepend=0., append=0.)

def diffus_step(y, args):
    return y + 0.1 * laplacian_numerator(y)

def loss(y, args):
    return diffus_step(y, args) - y

tridiag_solver = lineax.Tridiagonal()
newton_solver = optimistix.Newton(rtol=1e-8, atol=1e-8, linear_solver=lineax.Tridiagonal())

@jax.jit
def with_root_find(rhs):
    return optimistix.root_find(loss, newton_solver, rhs, tags=frozenset({lineax.tridiagonal_tag}), max_steps=1, throw=False)


@jax.jit
def jac_func_to_sol(rhs):
    jac = jax.jacfwd(diffus_step)(rhs, None)
    op = lineax.MatrixLinearOperator(jac, tags=frozenset({lineax.tridiagonal_tag}))
    return lineax.linear_solve(op, rhs, tridiag_solver, throw=False)


def tridiag_sparsity(x):
  dummy_tridiag = jnp.diag(x) + jnp.diag(x[:-1], 1) + jnp.diag(x[1:], -1)
  return jax.experimental.sparse.BCOO.fromdense(dummy_tridiag)

@jax.jit
def sparsejac_func_to_sol(rhs):
    with jax.ensure_compile_time_eval():
       sparsejac_func = sparsejac.jacfwd(diffus_step, sparsity=tridiag_sparsity(jnp.ones_like(rhs)))
    jac = sparsejac_func(rhs, None).todense()
    op = lineax.MatrixLinearOperator(jac, tags=frozenset({lineax.tridiagonal_tag}))
    return lineax.linear_solve(op, rhs, tridiag_solver, throw=False)

@jax.jit
def using_ones_like(rhs):
    du = jnp.ones_like(rhs[:-1])*0.1
    d = jnp.ones_like(rhs)*0.8
    d = d.at[0].set(0.9)
    d = d.at[-1].set(0.9)
    dl = jnp.ones_like(rhs[1:])*0.1
    op = lineax.TridiagonalLinearOperator(d, dl, du)
    return lineax.linear_solve(op, rhs, tridiag_solver, throw=False)

@jax.jit
def jlo_to_sol(rhs):
    op = lineax.JacobianLinearOperator(diffus_step, rhs, tags=frozenset({lineax.tridiagonal_tag}), jac="fwd")
    return lineax.linear_solve(op, rhs, tridiag_solver, throw=False)
```
</details>

Comparison for various sizes of `rhs` (using `EQX_ON_ERROR=nan` and `jax_enable_x64` with all other environment variables at their defaults) :
![image](https://github.com/user-attachments/assets/127c7e2d-c64f-4ef5-9a34-4ae5cdd0f068)

We see that this PR will lead to improvements in Newton solve (which uses `FunctionLinearOperator` under the hood) and `linear_solve` with `JacobianLinearOperator` for array sizes beyond 75. `JacobianLinearOperator` is now as efficient as `TridiagonalLinearOperator` (at least with a linear function).  A performance gain >15x is observed for array sizes of 1E3 and of greater 400x for array sizes of 1E4. The performance hit for smaller array sizes is probably again due to threading which I haven't controlled here (happy to look into it). As discussed in #149 the factor of 2 performance hit of the Newton solve is likely due to the Caucy termination test which requires two steps.

I have not done testing for custom primitives yet, let me know whether I should add this.

### Testing done
- [x] CI passes
- [x] Added ` test_tridiagonal` test in `test_operators.py`
- [x] Added `make_jacfwd_operator` and `make_jacrev_operator` in `helpers.py` to ensure this continues to work when `jac=True`

jpbrodrick89 committed:
Merge pull request #1 from patrick-kidger/main

merge main into fork

jpbrodrick89 committed:
Streamline diagonalisation by avoiding computation/instantiation of entire Jacobian matrix

jpbrodrick89 committed:
changed to using mv instead of jvp

jpbrodrick89 committed:
initialise test_diagonal  operators with diagonal matrix

jpbrodrick89 committed:
Merge branch 'patrick-kidger:main' into main

jpbrodrick89 committed:
remove redundant original diagonal overload for  FunctionLinearOperator

jpbrodrick89 committed:
Merge branch 'main' into jpb/diagfromjac

jpbrodrick89 committed:
Streamline tridiagonalisation of JacobianLinearOperator/FunctionLinearOperator using coloring methods

jpbrodrick89 committed:
add tests for JaobianLinearOperator with jac set to fwd/bwd

jpbrodrick89 committed:
fix jac=bwd + optimization

jpbrodrick89 committed:
skipped jacrev adjoint test for complex128

jpbrodrick89 committed:
skipped jacrev small_well_posed test for complex128 (possibly overkill: skipped 80 tests but only 4 failed)

jpbrodrick89 committed:
unravel instead of map

jpbrodrick89 committed:
continue to extract diagonal if no diagonal tag

jpbrodrick89 committed:
rmeove print statement in test_operator

Commit:
Decrease default value to prevent overflow in 32-bit.

Commit:
use maximum value of runtime dtype

johannahaffner commented:
Hi, just wanted to ask what the status is for this PR? We got a request for a new release, and I want to compile a To-Do list for what we want to have in before we do a release.

Commit:
limit condition numbers for sensitive, iterative solvers

Commit:
cond_cutoff -> 900 for GMRES, LSMR

Commit:
decrease condition number cutoff for LSMR

Commit:
implement safeguard for int overflow in 32 bit

Commit:
add extra safeguard for complex dtypes

adconner committed:
Replace allow_dependent_rows and allow_dependent_columns with assume_full_rank (#158)

The two functions allow_dependent_{rows,columns} together did the job of
answering if the solver accepts full rank matrices for the purposes of
the jvp. Allowing them to be implemented separately created some issues:
1) Invalid states were representable. Eg. What does it mean that
   dependent columns are allowed for square matrices if dependent rows
   are not? What does it mean that dependent rows are not allowed for
   matrices with more rows than columns?
2) As the functions accept operator as input, a custom solver could in
   principle decide its answer based on operator's dynamic value rather
   than only jax compilation static information regarding it, as in all
   the lineax defined solvers. This would prevent jax compilation and
   jit.

Both issues are addressed by asking the solver to report only if it
assumes the input is numerically full rank. If this assumption is
exactly violated, its behavior is allowed to be undefined, and is
allowed to error, produce NaN values, and produce invalid values.

patrick-kidger committed:
pyright fixes

adconner committed:
Implement Normal, a solver applying an inner solver to the normal equations (#159)

patrick-kidger committed:
Tidy up results. (#170)

patrick-kidger committed:
Update infra (#182)

Commit:
fix typo

Commit:
reset condition number check

Commit:
simplify dtype conversion: reuse function from our misc module

johannahaffner committed:
Merge pull request #176 from johannahaffner/lsmr-init-overflow

Decrease default value to prevent overflow in 32-bit.

patrick-kidger committed:
Moved bool states to Static.

There seem to be some spurious downstream failures in Diffrax with JAX 0.8.2 otherwise. Probably JAX has started promoting these to tracers on some unusual codepath.

jpbrodrick89 committed:
simplify jac=bwd behaviour

jpbrodrick89 committed:
Merge branch 'dev' into jpb/diagfromjac

jpbrodrick89 committed:
add PyTreeLinearOperator optimisations for mv and diagonal

jpbrodrick89 committed:
add type annotation to _leaf_from_keypath

jpbrodrick89 committed:
Merge branch 'main' into jpb/diagfromjac

jpbrodrick89 committed:
add efficient Matrix mv path, and Jacobian/Function materialise path

jpbrodrick89 committed:
add try sparse materialise helper (#3)

* Add sparse materialisation helper and efficient diagonal paths

This PR introduces _try_sparse_materialise helper and optimizes diagonal
operator handling throughout lineax.

Key changes:
- Add _try_sparse_materialise() that converts diagonal-tagged operators to
  DiagonalLinearOperator, preserving pytree structure via unravel
- Add efficient diagonal() for JLO/FLO using single JVP/VJP with ones basis
- Add efficient diagonal() for Composed: diag(A @ B) = diag(A) * diag(B)
- Simplify mv() for MLO, PTLO, Add, Composed to use _try_sparse_materialise
- Apply early sparse materialisation in materialise() registrations

Aux handling:
- Fix bug: linearise/materialise now preserve aux on AuxLinearOperator
- Preserve aux from first operator in Composed (output comes from op1)
- Inner aux in Add children silently stripped (unclear semantics - may
  warrant guards in future)

---------

Co-authored-by: jpbrodrick89 <jpbrodrick89@users.noreply.github.com>

jpbrodrick89 committed:
Merge branch 'main' into jpb/diagfromjac

jpbrodrick89 committed:
unify JLO and FLO diagonal registrations

jpbrodrick89 committed:
rename sparse variable -> maybe_sparse_op

jpbrodrick89 committed:
move and shorten construct diagonal basis

jpbrodrick89 committed:
revert unecessary **2.0 -> more efficient **2

jpbrodrick89 committed:
comments on why we can't use try sparse materialise with JLO

jpbrodrick89 committed:
Merge branch 'main' into jpb/diagfromjac

jpbrodrick89 committed:
out -> maybe_sparse_op

jpbrodrick89 committed:
merge main without pre-materialising IdentityLO

jpbrodrick89 committed:
remove duplicated AddLinearOperator registration

jpbrodrick89 committed:
deduplicate AddLinearOperator check registrations

jpbrodrick89 committed:
move leaf_from_keypath

jpbrodrick89 committed:
robust forward compatible diagonal extraction from ComposedLinearOperator

jpbrodrick89 committed:
restore to jpb/diagfromjac

jpbrodrick89 committed:
neatened up PR on top of diagfromjac and remove diagonal shadowing

jpbrodrick89 committed:
Merge branch 'main' into jpb/tridiagfromjac

jpbrodrick89 committed:
only sparse materialise to tridiagonal if in/out structure flat

jpbrodrick89 committed:
Merge branch 'main' into jpb/diagfromjac

jpbrodrick89 committed:
Merge branch 'main' into jpb/tridiagfromjac

jpbrodrick89 committed:
Merge branch 'main' into jpb/diagfromjac

jpbrodrick89 committed:
Merge branch 'jpb/diagfromjac' into jpb/tridiagfromjac

jpbrodrick89 committed:
Merge branch 'main' into jpb/tridiagfromjac

jpbrodrick89 committed:
add gather flags for performance robustness
